### PR TITLE
Sprint 6 (dev->stg)

### PR DESCRIPTION
**Pull request by afrasier**
This PR includes:

- Capsule Description CRUD endpoints #95 
- CSV Downloads #96 
- Dynamic field limiting #97 
- Notification system + saved search counts and notification creation #100 

For more details on each subgroup, see the specified PR

### CODE REVIEW

File: talentmap_api/common/common_helpers.py [hunk 1]
```diff
@@ -53,3 +55,55 @@ def validate_filters_exist(filter_list, filter_class):
...
+    the filter class into thinking we have passed a valid request. The steps to
+    accomplish this are the following:
+        1. Turn the filters into a MultiValueDict - this is required later for
+           the construction of our QueryDict, to prevent nested lists
+        2. Create a QueryDict of the filters - we need this to support some function
+           calls that are made internally which would error out on a normal dict
+        3. Create a fake request object, and set the QueryDict as the query_params
+        4. Get the subset filter class
+        5. Instantiate the subset filter class using query_params and the faked request
+        6. Get the queryset from the filter class
```
Comment (jseppi) [inline talentmap_api/common/common_helpers.py]: Nice comments 👍

File: talentmap_api/common/common_helpers.py [hunk 2]
```diff
@@ -53,3 +55,55 @@ def validate_filters_exist(filter_list, filter_class):
...
+    for key, value in filters.items():
+        if isinstance(value, list):
+            new_filters.setlist(key, value)
+        else:
+            new_filters.appendlist(key, value)
+
+    query_params = QueryDict('', mutable=True)
+    query_params.update(new_filters)
+
+    # Your daily dose of python wizardry: https://docs.python.org/3/library/functions.html#type
```
Comment (jseppi) [inline talentmap_api/common/common_helpers.py]: ✨✨✨

File: talentmap_api/user_profile/models.py [hunk 3]
```diff
@@ -79,11 +121,24 @@ def create_profile(sender, instance, created, **kwargs):
...
+    This listener ensures newly created or edited saved searches update their counts
+    '''
+    instance.update_count()
+
+
 @receiver(post_save, sender=Sharable)
-def email_notification(sender, instance, created, **kwargs):
+def post_sharable_save(sender, instance, created, **kwargs):
     '''
     This listener e-mails the receiving user to notify them of their share.
```
Comment (jseppi) [inline talentmap_api/user_profile/models.py]: I guess a more generic "notifies" might be more correct than "e-mails" now.

File: talentmap_api/common/mixins.py [hunk 4]
```diff
@@ -5,3 +5,26 @@ class ActionDependentSerializerMixin(object):
 
     def get_serializer_class(self):
             return self.serializers.get(self.action, self.serializers['default'])
+
+
+class FieldLimitableSerializerMixin(object):
+    '''
+    Supports limiting the return fields via the "fields" and "fields_exclude"
```
Comment (jseppi) [inline talentmap_api/common/mixins.py]: should this be `via the "include" and "exclude" ...`?

File: talentmap_api/common/serializers.py [hunk 5]
```diff
@@ -56,6 +63,24 @@ def __init__(self, *args, **kwargs):
...
     @classmethod
+    def correct_include_hierarchy(cls, override_fields):
+        '''
+        This method ensures that the list of override fields includes any items' parents
+        For example:
+            ["description__id"] would not render any data, as it is excluding the parent, "description"
+        This method fixes this issue by ensuring all parents are included
+
+        Args:
+            override_fields (list) - List of overriden fields to include
```
Comment (jseppi) [inline talentmap_api/common/serializers.py]: `overiden` -> `overridden`

File: talentmap_api/messaging/models.py [hunk 6]
```diff
@@ -0,0 +1,19 @@
...
+
+class Notification(models.Model):
+    '''
+    This model represents an individual notification item
+    '''
+
+    message = models.TextField(null=False, help_text="The message for the notification")
+    owner = models.ForeignKey('user_profile.UserProfile', on_delete=models.CASCADE, null=False, related_name="notifications", help_text="The owner of the notification")
+
+    read = models.BooleanField(default=False, help_text="Whether this notification has been read")
```
[the remark this answers is not in the document]
Comment (jseppi) [reply talentmap_api/messaging/models.py]: bleh, `has_been_read` is kinda ugly though :\

File: talentmap_api/position/views.py [hunk 7]
```diff
@@ -1,16 +1,20 @@
...
+from rest_framework.permissions import IsAuthenticated, IsAuthenticatedOrReadOnly
 
-from talentmap_api.position.models import Position, Grade, Skill
-from talentmap_api.position.filters import PositionFilter, GradeFilter, SkillFilter
-from talentmap_api.position.serializers import PositionSerializer, GradeSerializer, SkillSerializer
+from talentmap_api.common.mixins import FieldLimitableSerializerMixin
+
+from talentmap_api.position.models import Position, Grade, Skill, CapsuleDescription
+from talentmap_api.position.filters import PositionFilter, GradeFilter, SkillFilter, CapsuleDescriptionFilter
+from talentmap_api.position.serializers import PositionSerializer, GradeSerializer, SkillSerializer, CapsuleDescriptionSerializer
```
Comment (jseppi) [inline talentmap_api/position/views.py]: I think some of these `import`s can also be cleaned up

File: talentmap_api/messaging/models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (afrasier) [reply talentmap_api/messaging/models.py]: `is_read`?

File: talentmap_api/messaging/views.py [hunk 8]
```diff
@@ -0,0 +1,39 @@
+from rest_framework.viewsets import ReadOnlyModelViewSet, GenericViewSet
+from rest_framework import mixins
+from django.db.models import Q
+
+from rest_framework.permissions import IsAuthenticated, IsAuthenticatedOrReadOnly
```
[the remark this answers is not in the document]
Comment (afrasier) [reply talentmap_api/messaging/views.py]: This is true. Will do.

File: talentmap_api/common/common_helpers.py (hunk 1 above)
Comment (afrasier) [reply talentmap_api/common/common_helpers.py]: Mainly for me, since it took me a while to deduce. :)

File: talentmap_api/common/mixins.py (hunk 4 above)
Comment (afrasier) [reply talentmap_api/common/mixins.py]: Yep, will fix.

File: talentmap_api/messaging/models.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (jseppi) [reply talentmap_api/messaging/models.py]: haha, yeah, `is_read` works I think :)

File: talentmap_api/position/views.py (hunk 7 above)
Comment (afrasier) [reply talentmap_api/position/views.py]: Upgraded my linter to flag these automatically for me.

**Pull request merged**